### ISSUE TITLE
A generic `deepmerge` function

### ISSUE THREAD

**Issue by nitrocode** [opentofu/opentofu#790]
### OpenTofu Version

latest

### Use Cases

deep merge yaml without an external module

### Attempted Solutions

https://registry.terraform.io/modules/cloudposse/config/yaml/0.2.0/submodules/deepmerge

### Proposal

A generic `deepmerge` function

### References

- cc: @schollii for the original request in https://github.com/hashicorp/terraform/issues/31815

**Comment by cube2222** [comment]
Hey, thanks for submitting this issue @nitrocode!

What's the advantage of a built-in over using the external module?

**Comment by janosdebugs** [comment]
Hey @nitrocode thank you for this issue. We have discussed this issue in the core team and feel that deep merging data structures is a non-trivial issues, especially as it relates to `sensitive` values. We are generally open to accepting this issue, but we'd like to know the specifics of how this function should be have? This can take the form of an RFC or just explaining it in a comment.

**Comment by voidtek** [comment]
Since April 2024, no solution has been released in a stable version.
The idea of having a provider-defined function is logical in theory, but in reality, we are using tools that have not reached a stable release, and we have been in this situation for almost a year. This feature has not found a community to create a provider, module, or other implementation.

Moreover, given that the community can use the Go mergo library, implementing and maintaining a deep-merge function/feature within OpenTofu would be simple.

As of today, I spend a significant amount of time working around this limitation and making my code more complex because there is no straightforward deep-merge function that I can use in a foreach without relying on dynamic blocks, external data, or a custom module.

I would appreciate it if you could reopen this issue/conversation and reconsider integrating this feature.

**Comment by ollevche** [comment]
Re-opening this issue to discuss with the core team once again!

**Comment by apparentlymart** [comment]
I spent a couple hours writing a relatively-simple "deep merge" function that deals with the simple case of merging objects together recursively in the provider [`apparentlymart/deepmerge`](https://search.opentofu.org/provider/apparentlymart/deepmerge/latest). I implemented this directly using the `cty` type system that OpenTofu is built from, for (hopefully) minimal surprises.

Someone else previously wrote [`isometry/deepmerge`](https://search.opentofu.org/provider/isometry/deepmerge/latest) which offers a function that wraps the aforementioned "mergo" library. Since it uses mergo it has to first convert from OpenTofu types into Go types, perform the deep merge, and then convert back. Therefore its behaviors _could_ potentially include some Go-flavored surprises (I'm not sure!) but it offers some additional options for customizing the merge behavior in case the simple behaviors in my provider are insufficient for certain cases.

Given that, we do seem to have at least two providers available offering something like the requested functionality.

**Comment by ollevche** [comment]
We discussed this issue with the core team and we think the existing community-maintained solutions should cover the use-case above. Thanks @apparentlymart for providing one!

We are happy to re-open the discussion if the context changes. Alternatively, you could report issues to respective provider authors.

**Comment by voidtek** [comment]
I'm sorry to insist, but we seem to have different definitions of "community-maintained."

The package 'isometry/deepmerge' is at version 0.4.2 and has been around for over a year. Meanwhile, 'apparentlymart/deepmerge' is at version 1.0.0, but the repository was created just yesterday and lacks continuous integration.

I don't mean to be unpleasant, and I respect the work done, but these are neither reliable nor trustworthy solutions for professional use.

I also don’t understand the argument that "there are community-maintained solutions" as a reason to avoid adding a simple function—especially for deep merging, considering how much time we spend manipulating objects in our infrastructures as code.

As @apparentlymart  suggested, we could use 'cty' to minimize surprises.

**Comment by apparentlymart** [comment]
Hi @yordis. Can you say more about what you would consider the advantage of putting this provider in an "opentofu-contrib" organization?

The OpenTofu Registry is intended to be the place where community-driven work is published. _Just_ putting the provider in a different namespace would not change anything about how it is maintained or published, so I assume you must be imagining that GitHub organization being treated in some different way than other GitHub accounts/organizations, and so I'd just like to understand exactly what your expectations are for it.

Thanks!

**Comment by yordis** [comment]
> Hi @yordis. Can you say more about what you would consider the advantage of putting this provider in an "opentofu-contrib" organization?

I am most likely thinking about `crossplane-contrib` organization when I said that, primarily, processes, expectations and so on that does not requires the "Core" maintainers to take full ownership of, but creates an environment where promotes some alignment and have some level of "recommended" packages since the process itself makes sure that that "reserved" namespace comes with responsibilities and level of maturity.

Or amybe the registries add tags such as "recommended", some level of curation. I understand that it may cause more work for the "Core" team or the registry team, but such level of curation could be offload, or, in the worst case, be extremely conservative.

**Comment by apparentlymart** [comment]
Thanks for that extra context, @yordis. I'm understanding that your primary interest is in having a team of curators that identifies community projects that meet some sort of documented standard.

That seems like a broader question than whether an individual function should be offered in a provider or built in, so we should probably discuss that idea in a separate issue. Would you be willing to create one? It would help if you could mention exactly which criteria you'd like this team of curators to prioritize. As you noted, there are multiple ways the curators could share their work -- including a special GitHub org or some sort of labeling on the registry -- but it feels to me like the primary question is what exactly this hypothetical curation team would be looking for in the projects they choose to list, while the mechanism of sharing that result is a secondary detail.

Thanks again!

**Comment by yordis** [comment]
@apparentlymart https://github.com/orgs/opentofu/discussions/2615

**Comment by voidtek** [comment]
I completely agree with @yordis. To emphasize the necessity of having a part of this under the umbrella of an organization, we can see that one of Deepmerge's function-based solutions is managed by a single person, and SemVer is not a priority - [https://github.com/isometry/terraform-provider-deepmerge/issues/75](https://github.com/isometry/terraform-provider-deepmerge/issues/75#issuecomment-2738481568)
From a business perspective, this does not inspire confidence for the long term, and we don’t want to end up maintaining a fork of our project.

**Comment by isometry** [comment]
As the author of [isometry/deepmerge](https://github.com/isometry/terraform-provider-deepmerge), let me say that I take backward compatibility very seriously; I just hadn't attached too much mysticism to "v1.0". Inspired by @voidtek's desire for such, and the fact that I'm approaching 1M downloads from the [Terraform Registry](https://registry.terraform.io/providers/isometry/deepmerge/latest), and I have no idea how many more through the [OpenTofu Registry](https://search.opentofu.org/provider/isometry/deepmerge/latest), I've just released [v1.0.0](https://github.com/isometry/terraform-provider-deepmerge/releases/tag/v1.0.0). Happy merging.
